File: src/main/java/com/github/squirrelgrip/meter/exception/MeterReaderException.java
package com.github.squirrelgrip.meter.exception;

import com.github.squirrelgrip.meter.domain.ContextKey;

import java.util.Map;

public class MeterReaderException extends RuntimeException {
    private final Map<ContextKey, Object> context;

    public MeterReaderException(String message, Map<ContextKey, Object> context) {
        this(message, context, null);
    }

    public MeterReaderException(String message, Map<ContextKey, Object> context, Exception cause) {
        super(message, cause);
        this.context = Map.copyOf(context); // Make sure we copy the current context as Immutable
    }

    public Map<ContextKey, Object> getContext() {
        return context;
    }
}

File: src/main/kotlin/com/github/squirrelgrip/meter/exception/MeterReaderException.kt
package com.github.squirrelgrip.meter.exception

import com.github.squirrelgrip.meter.domain.BaseMeterContext
import com.github.squirrelgrip.meter.domain.MeterContext

open class MeterReaderException(
    message: String,
    meterContext: MeterContext,
    cause: Exception? = null
) : RuntimeException(message, cause) {
    val context: MeterContext = meterContext.clone()
}
